src/communication/xcConnection.ts

```diff
@@ -6,90 +6,72 @@ let log = require("loglevel");
+    getXcApiList(serverUrl: string): Promise<Array<String>> {
+        const session = SessionFactory(serverUrl, null, null);
+        return session.init()
+            .then(_ => {
+                return session.privateSubscriber.getXcApiList();
+            })
+            .catch(err => {
+                log.debug("getModel request failed");
+                log.debug(err);
             });
-        };
-        let errorListener = (err: Error) => {
-            getXcApiListListener(err, null);
-            log.debug("Error while getting Apis List");
-            log.debug(err);
-        };
-        let closeListener = (_: CloseEvent) => {
-        };
-        session.init(openListener, errorListener, closeListener);
     };
 
-    createSession(xcApiFileName: string, serverUrl: string, createSessionListener: (error: Error, session: Session) => void, disconnectionErrorListener: (closeEvent: CloseEvent) => void): void {
-        this.init(xcApiFileName, serverUrl, null, createSessionListener, disconnectionErrorListener);
+    createSession(xcApiFileName: string, serverUrl: string): Promise<Session> {
+        return this.init(xcApiFileName, serverUrl, null);
     };
 
-    createAuthenticatedSession(xcApiFileName: string, serverUrl: string, sessionData: string, createAuthenticatedSessionListener: (error: Error, session: Session) => void, disconnectionErrorListener: (closeEvent: CloseEvent) => void): void {
-        this.init(xcApiFileName, serverUrl, sessionData, createAuthenticatedSessionListener, disconnectionErrorListener);
+    createAuthenticatedSession(xcApiFileName: string, serverUrl: string, sessionData: string): Promise<Session> {
+        return this.init(xcApiFileName, serverUrl, sessionData);
     };
 
-    private init(xcApiFileName: string, serverUrl: string, sessionData: string, createSessionListener: (error: Error, session: Session) => void, disconnectionErrorListener: (closeEvent: CloseEvent) => void): void {
-        let session = SessionFactory(serverUrl, null, sessionData);
-        let thisConnection = this;
-        let getXcApiRequest = (xcApiFileName: string, createSessionListener: (error: Error, session: Session) => void) => {
-            session.privateSubscriber.getXcApi(xcApiFileName, (xcApi: string) => {
-                if (xcApi != null) {
-                    const parser = new DefaultApiConfigurationParser();
-                    const configurationPromise = parser.parse(xcApi);
-                    configurationPromise.then(configuration => {
-                        session.configuration = configuration;
-                        session.replyPublisher.configuration = configuration;
-                        createSessionListener(null, session);
-                    }).catch(e => createSessionListener(e, null));
-                } else {
-                    createSessionListener(new Error(`Unknown Api: ${xcApiFileName}`), null);
+    getUnexpectedCloseSessionError(serverUrl: string): Promise<Error> {
```

Comment:
It is not a good idea to create a new session to check unexpected errors.
Just add a method called UnexpectedErrors to the Session object. In this case provide a callback instead of a promise (I think it is the only case in our code where callbacks are better than promises)

Reply:
We don't really need a function directly the api object (it is related to the session...)

Reply:
We need both.

For me the best approach is that when you create a session you pass an (optional) callback for when the session is unexpected close and you get a promise of your session

Reply:
I think it is better to put the method in the session object. It is strange to have a promise and a callback in the same method

Reply:
The problem is, what if you loose connection between getting your promise and calling the callback object?

I think we should provide a callback as early as possible (even before the session is created)

Reply:
@maurelio1234 therefore it is not an error callback related to the session, but a global error handler ? @mbarekh